### PR TITLE
fix: thing: Update schema message flow and error forwarding

### PR DESCRIPTION
**Pull request by lcbm**
**What does this PR introduce?**

This PR fixed a bug where the Update Schema message flow was incorrect, as well as a 
bug which Update Schema didn't forward errors to the client (as is specified here CESARBR/knot-fog-connector/tree/v4.0.0/docs/api/amqp.md). To achieve this, interactors were modified (and consequently, their tests and mocks) to return an error (instead of a string pointer) -- moving the responsibility of retrieving a pointer to the error's message to `pkg/thing/delivery/amqp/client.go`.

**What kind of change does this PR introduce?** (check at least one)

- [X] Bug fix
- [ ] Feature
- [ ] Code style update
- [X] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce any breaking changes?** (check one)

- [ ] Yes
- [X] No

If yes, please describe the impact and migration path for existing applications:

**Does this reference any currently open issues?**
(use the [special GitHub keywords](https://help.github.com/en/articles/closing-issues-using-keywords) to reference issues)

Nope :grimacing: 

**Any relevant logs, error output, etc?**

Nope :smile: 

**Any additional comments?**

Nope :heart_eyes: 

**Where has this been tested?**

- Operating System/Platform: Linux 5.5.9-arch1-2 x86_64 GNU/Linux
- Go Version: go1.14 linux/amd64

### CODE REVIEW

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/thing/interactors/update_schema.go  1
- pkg/thing/interactors/update_schema_test.go  5
- pkg/thing/interactors/register_thing.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5206347&bid=17075570)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/thing/interactors/unregister_thing_test.go  2
         

Clones removed
==============
+ pkg/thing/interactors/update_schema_test.go  -9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5206347&bid=17075570)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/thing/interactors/unregister_thing_test.go  2
         

Clones removed
==============
+ pkg/thing/interactors/update_schema_test.go  -9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5206347&bid=17075570)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/thing/interactors/unregister_thing_test.go  2
         

Clones removed
==============
+ pkg/thing/interactors/update_schema_test.go  -9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5206347&bid=17075570)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/thing/interactors/unregister_thing_test.go  2
         

Clones removed
==============
+ pkg/thing/interactors/update_schema_test.go  -9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5206347&bid=17075570)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/thing/interactors/unregister_thing_test.go  2
         

Clones removed
==============
+ pkg/thing/interactors/update_schema_test.go  -9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5206347&bid=17075570)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/thing/interactors/unregister_thing_test.go  2
         

Clones removed
==============
+ pkg/thing/interactors/update_schema_test.go  -9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5206347&bid=17075570)

File: pkg/thing/interactors/update_schema.go [hunk 1]
```diff
@@ -7,12 +7,10 @@ import (
...
 
-// ErrInvalidSchema represents the error when the schema has a invalid format
-type ErrInvalidSchema struct{}
-
-func (eis *ErrInvalidSchema) Error() string {
-	return "Thing's schema is invalid"
-}
+var (
+	// ErrSchemaInvalid is returned for invalid schema formats.
+	ErrSchemaInvalid = fmt.Errorf("invalid schema")
```
Comment (netoax) [inline pkg/thing/interactors/update_schema.go]: `fmt.Errorf` should be used only when it's necessary to fill some placeholders with additional data. Please, change to `errors.New()`.

**Review by netoax — CHANGES_REQUESTED** [review]
@lcbm your patch is very good, thanks!!! 😄 
I'd just suggest you break the last commit in smaller ones.

I can see three main changes you did:

- enable UpdateSchema to send error
- move error’s string extraction to delivery layer
- error improvements, example:

```
- // ErrInvalidSchema represents the error when the schema has an invalid format
- type ErrInvalidSchema struct{}

- func (eis *ErrInvalidSchema) Error() string {
- 	return "Thing's schema is invalid"
- }

+ var (
+	// ErrSchemaInvalid is returned for invalid schema formats.
+	ErrSchemaInvalid = fmt.Errorf("invalid schema")
+)
```

File: pkg/thing/interactors/update_schema.go (hunk 1 above)
Comment (lcbm) [reply pkg/thing/interactors/update_schema.go]: Done :smile:

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/thing/interactors/unregister_thing_test.go  2
         

Clones removed
==============
+ pkg/thing/interactors/update_schema_test.go  -9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5206347&bid=17075570)

**Review by esdrasjnr — APPROVED** [review]
LGTM! Excellent patch, thanks @lcbm :smiley:

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/thing/interactors/unregister_thing_test.go  2
         

Clones removed
==============
+ pkg/thing/interactors/update_schema_test.go  -9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5206347&bid=17075570)

**Comment by netoax** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/thing/interactors/unregister_thing_test.go  2
         

Clones removed
==============
+ pkg/thing/interactors/update_schema_test.go  -9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5206347&bid=17075570)

**Review by netoax — APPROVED** [review]

**Pull request merged**